google/cloud/internal/opentelemetry_context.cc

```diff
@@ -0,0 +1,80 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef GOOGLE_CLOUD_CPP_HAVE_OPENTELEMETRY
+#include "google/cloud/internal/opentelemetry_context.h"
+#include <opentelemetry/context/runtime_context.h>
+#include <opentelemetry/nostd/unique_ptr.h>
+
+namespace google {
+namespace cloud {
+GOOGLE_CLOUD_CPP_INLINE_NAMESPACE_BEGIN
+namespace internal {
+namespace {
+
+using OTelContextWithTokens = std::vector<
+    std::pair<opentelemetry::context::Context,
+              opentelemetry::nostd::unique_ptr<opentelemetry::context::Token>>>;
+
+/**
+ * Holds a context stack to be used when tracing asynchronous APIs.
+ *
+ * Note that opentelemetry::context::Context is a pimpl class. Copying it is
+ * cheap.
+ */
+OTelContextWithTokens& ThreadLocalOTelContext() {
+  thread_local auto context = OTelContextWithTokens{};
+  return context;
+}
+
+}  // namespace
+
+OTelContext CurrentOTelContext() {
+  auto const& v = ThreadLocalOTelContext();
+  OTelContext contexts;
+  contexts.reserve(v.size());
+  std::transform(v.begin(), v.end(), std::back_inserter(contexts),
+                 [](auto const& p) { return p.first; });
+  return contexts;
+}
+
+void PushOTelContext() {
+  auto& v = ThreadLocalOTelContext();
+  v.emplace_back(opentelemetry::context::RuntimeContext::GetCurrent(), nullptr);
+}
+
+void PopOTelContext() {
+  auto current = opentelemetry::context::RuntimeContext::GetCurrent();
+  DetachOTelContext(current);
+}
+
+void AttachOTelContext(opentelemetry::context::Context const& context) {
+  auto& v = ThreadLocalOTelContext();
+  auto token = opentelemetry::context::RuntimeContext::Attach(context);
+  v.emplace_back(context, std::move(token));
+}
+
+void DetachOTelContext(opentelemetry::context::Context const& context) {
+  auto& v = ThreadLocalOTelContext();
+  auto it = std::find_if(v.begin(), v.end(), [&context](auto const& p) {
```

Comment:
On second thought... maybe we should keep these in a multiset or some other data structure where `std::find()` can run faster?

Even a `std::list` would be better as the `.erase()` for vector is $O(n)$, whereas `std::list::erase` is $O(1)$.  You still would have a $O(n)$ search.. that is where using `std::multiset<>` would give you $O(ln(n))$

Reply:
We would only see duplicated `Context` objects in the list in practice when the futures are executed immediately. In that case we would basically take the current stack and duplicate it.

I think this was a tad lazy, so now we will check if we are in the same thread, and if so, we will not touch the stack.

---

Re: time complexity....

Now that the `Context` objects in the stack are all unique, and we can only compare equal, we cannot beat $O(n)$ search. (Also, I will remind that `n <= 5`).

Updated to use a `std::list` for the better erase.

Reply:
I see `v.erase(it, v.end())`, which seems to be O(n) no matter what container you use.

Reply:
If you are erasing from the back, do you need to use find anyway?

This is what I thought when I first looked at the doc

deque<Context> stack;

while(stack.back() != c) {
  stack.pop_back();
}

(no need to make changes, just thought I'd leave my thoughts)

Reply:
> If you are erasing from the back, do you need to use find anyway?
>
> This is what I thought when I first looked at the doc
> ```
> deque stack;
>
> while(stack.back() != c) {
> stack.pop_back();
> }
> ```

Yeah, that is true, and how I was thinking about it. Your API might be better.

The OTel `Detach()` implementation will find the element in their stack (searching from the back) and pop every element along the way (assuming the element is in the stack). I was copying that, but it was probably over cautious.

I think we can assume that the thing we are trying to pop is either `back()`, or it isn't found in the stack. I don't think we would leak memory.

Reply:
I think your impl is good. What if it doesn't find it on the stack? My suggestion would delete the whole stack anyway

Reply:
If you think it is either `back()` or not there the code would be simpler (don't forget the empty collection case).

Otherwise, consider `std::find_if(stack.rbegin(), stack.rend(), ...);`